[ClangImporter] Take isCompatibilityAlias() into account in interface printing

rintaro opened the pull request:
If the Clang declarations are *types*, canonical declaration in Swift is imported for newest version of Swift. In interface generation, if the declaration is versioned and it's imported as a member, we might have to search the canonical imported decl in newest version of Swift.

rdar://problem/39295365

rintaro committed:
[ClangImporter] Take isCompatibilityAlias() into account in interface printing

If the Clang declrations are *types*, canonical declaration in Swift is
imported for newest version of Swift. In interface generation, if the
declaration is versioned and it's imported as a member in either or both
version of Swift, we have to take compatibility typealias into account.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@jrose-apple I updated the PR to use `TypeAliasDecl::isCompatibilityAlias`. How does this look?
@swift-ci Please smoke test

jrose-apple commented on test/SourceKit/InterfaceGen/gen_clang_module.swift.apinotes_swift3.response:
```diff
@@ -0,0 +1,1232 @@
+}
+extension MemberToMember_Class_Swift3 {
+
+    open class Payload : NSObject {
```
This is incorrect. I would expect a typedef here that points to the real name that ended up getting used, MemberToMember_Class_Swift4.Payload. (And the same for the other member-to-member test).

rintaro replied:
Yeah, I agree. I expect:
```swift
extension MemberToMember_Class_Swift3 {
    typealias Payload = MemberToMember_Class_Swift4.Payload
}
extension MemberToMember_Class_Swift4 {
    open class Payload : NSObject {
    }
}
```
I'm investigating. But I think it's possible that ImportDecl doesn't do the right thing, because my change does not  manipulate the AST hierarchy.

jrose-apple replied:
Yep, definitely possible!

jrose-apple reviewed (commented):
Hm, I don't think this is the right logic yet for the member-to-member case (see comment).

rintaro committed:
[ClangImporter] Fixed interface printing for MemberToMember migration

rintaro reviewed (commented):
Partially fixed member-to-member pattern. But still have  a problem.

rintaro commented:
@swift-ci Please smoke test

jrose-apple commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -2476,29 +2476,28 @@ void ClangModuleUnit::getTopLevelDecls(SmallVectorImpl<Decl*> &results) const {
-      }
-      if (!ext) continue;
+      // If this is a compatibility typealias, the canonical extension may
+      // exists in another extension.
```
I think this is supposed to say "the canonical *type* (declaration) may exist in another extension".

jrose-apple commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2175,7 +2175,9 @@ namespace {
-              activeName.getDeclName() == canonicalName.getDeclName()) {
+              activeName.getDeclName() == canonicalName.getDeclName() &&
+              activeName.getEffectiveContext().equalsWithoutResolving(
+                  canonicalName.getEffectiveContext())) {
```
So it was an importer bug after all!

jrose-apple commented on test/SourceKit/InterfaceGen/gen_clang_module.swift.apinotes_swift3.response:
```diff
@@ -0,0 +1,2341 @@
+}
+extension MemberToMember_SameName_Class_Swift3 {
+
+    public typealias Payload = MemberToMember_SameName_Class_Swift4.Payload
```
Oh hey, there's even a FIXME about it. :-( https://github.com/apple/swift/blob/master/lib/ClangImporter/SwiftLookupTable.cpp#L1654-L1655

I'm not sure if it would be fully correct to check `EffectiveClangContext::equalsWithoutResolving` there as well, but it's probably still *more* correct than what's there today.

rintaro replied:
Thank you!
I reused `ClangImporter::Implementation::forEachDistinctName`.

rintaro committed:
[ClangImporter] Moved implementation of forEachDistinctName to NameImporter

Renamed to NameImpoter::forEachDistinctImportName.

And fixed importer::addEntryToLookupTable to use it so that it can
distinct same names in different contexts.

rintaro committed:
[ClangImporter] Fix comment

rintaro commented:
@swift-ci Please smoke test

rintaro commented on lib/ClangImporter/SwiftLookupTable.cpp:
```diff
@@ -1633,37 +1633,30 @@ void importer::addEntryToLookupTable(SwiftLookupTable &table,
+        if (version == currentVersion && importedName.isSubscriptAccessor()) {
+          table.addEntry(DeclName(nameImporter.getContext(),
+                                  DeclBaseName::createSubscript(),
+                                  ArrayRef<Identifier>()),
```
I'm wondering this should be `{ Identifier() }` like https://github.com/rintaro/swift/blob/f2085672/lib/ClangImporter/ImportDecl.cpp#L6692

jrose-apple replied:
Oops, yes, it should be.

rintaro committed:
[ClangImporter] Add member-to-member versioned type test

rintaro commented:
@swift-ci Please smoke test

jrose-apple reviewed (approved):
That was more of an adventure than I expected. Nice work!

rintaro committed:
[ClangImporter] Fix declname for subscript in SwiftLookupTable

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
Thank you @jrose-apple for supporting! I learned a lot about `ClangImporter` in this PR 😄
